### PR TITLE
Stun batons cannot be turned on for 8 Seconds after being dropped

### PR DESCRIPTION
**Pull request by JamieD1**
Stun batons cannot be turned on for 8 Seconds after being dropped

Theos half approved

Most likely needed after 1 hit might be merged.

:cl:  
rscadd: Stun batons cannot be turned on for 8 Seconds after being dropped
/:cl:

### CODE REVIEW

File: code/game/objects/items/stunbaton.dm [hunk 1]
```diff
@@ -15,7 +15,7 @@
 	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 50, BIO = 0, RAD = 0, FIRE = 80, ACID = 80)
 
 	var/cooldown_check = 0
-
+	var/dropcheck = FALSE
```
Comment (SomeguyManperson) [inline code/game/objects/items/stunbaton.dm:19]:
```suggestion
	/// if the baton is on cooldown from being  dropped
	var/dropcheck = FALSE
```

File: code/game/objects/items/stunbaton.dm [hunk 2]
```diff
@@ -58,6 +58,16 @@
 	if(status && prob(throw_hit_chance) && iscarbon(hit_atom))
 		baton_stun(hit_atom)
 
+/obj/item/melee/baton/dropped(mob/user, silent)
+	. = ..()
+	status = FALSE
+	dropcheck = TRUE
```
Comment (SomeguyManperson) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
	dropcheck = TRUE
	visible_message(span_warning("The safety strap on [src] is pulled as it is dropped, triggering its emergency shutoff!"))
```

File: code/game/objects/items/stunbaton.dm [hunk 3]
```diff
@@ -124,6 +134,8 @@
 		return ..()
 
 /obj/item/melee/baton/attack_self(mob/user)
+	if(dropcheck == TRUE)
+		to_chat("Baton has been dropped and needs to recalibrate!")
```
Comment (SomeguyManperson) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
		to_chat("[src]'s emergency shutoff is still active!")
```

File: code/game/objects/items/stunbaton.dm [hunk 4]
```diff
@@ -58,6 +58,16 @@
...
 
+/obj/item/melee/baton/dropped(mob/user, silent)
+	. = ..()
+	status = FALSE
+	dropcheck = TRUE
+	addtimer(CALLBACK(src, .proc/reclibrate), 8 SECONDS)
+
+/obj/item/baton/proc/recalibrate()
+	dropcheck = FALSE
+	status = TRUE
```
Comment (SomeguyManperson) [inline code/game/objects/items/stunbaton.dm]: make it play some sound here or something to show it's on again

File: code/game/objects/items/stunbaton.dm [hunk 5]
```diff
@@ -124,6 +134,8 @@
 		return ..()
 
 /obj/item/melee/baton/attack_self(mob/user)
+	if(dropcheck == TRUE)
```
Comment (SomeguyManperson) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
	if(dropcheck)
```

File: code/game/objects/items/stunbaton.dm [hunk 6]
```diff
@@ -58,6 +58,16 @@
 	if(status && prob(throw_hit_chance) && iscarbon(hit_atom))
 		baton_stun(hit_atom)
 
+/obj/item/melee/baton/dropped(mob/user, silent)
+	. = ..()
+	status = FALSE
+	dropcheck = TRUE
+	addtimer(CALLBACK(src, .proc/recalibrate), 8 SECONDS)
+
+/obj/item/baton/proc/recalibrate()
```
Comment (SomeguyManperson) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
/obj/item/melee/baton/proc/recalibrate()
```

**Comment by maxion12345** [comment]
heres an idea

what if we just didn't merge one-hit batons so all of this wouldn't be needed

**Comment by ghost** [comment]
> heres an idea
> 
> what if we just didn't merge one-hit batons so all of this wouldn't be needed

Based

File: code/game/objects/items/stunbaton.dm [hunk 7]
```diff
@@ -58,6 +59,21 @@
 	if(status && prob(throw_hit_chance) && iscarbon(hit_atom))
 		baton_stun(hit_atom)
 
+/obj/item/melee/baton/dropped(mob/user, silent)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm:68]: update icon

**Comment by Mqiib** [comment]
Also gives you an 8 second timer before it can be turned on when you pull it out of a belt, which is less than ideal. please fix (also dont merge 1-hit batong pls)

**Comment by JamieD1** [comment]
> Also gives you an 8 second timer before it can be turned on when you pull it out of a belt, which is less than ideal. please fix (also dont merge 1-hit batong pls)

Fun fact it was fixed before you made this comment

File: code/__DEFINES/components.dm [hunk 8]
```diff
@@ -156,6 +156,7 @@
 #define COMSIG_MOVABLE_BUCKLE "buckle"							//from base of atom/movable/buckle_mob(): (mob, force)
 #define COMSIG_MOVABLE_UNBUCKLE "unbuckle"						//from base of atom/movable/unbuckle_mob(): (mob, force)
 #define COMSIG_MOVABLE_PRE_THROW "movable_pre_throw"			//from base of atom/movable/throw_at(): (list/args)
+#define COMSIG_MOVABLE_PRE_DROPTHROW "movable_pre_dropthrow"	//from base of atom/movable/throw_at(): (list/args)
```
Comment (ToasterBiome) [inline code/__DEFINES/components.dm]: explain what this does since they're both from base throw_at

File: code/game/objects/items/stunbaton.dm [hunk 9]
```diff
@@ -51,6 +52,7 @@
 			log_mapping("[src] at [AREACOORD(src)] had an invalid preload_cell_type: [preload_cell_type].")
 		else
 			cell = new preload_cell_type(src)
+	RegisterSignal(src, COMSIG_MOVABLE_PRE_DROPTHROW, .proc/throwbaton)
```
Comment (ToasterBiome) [inline code/game/objects/items/stunbaton.dm:55]: You can't use VARSET_CALLBACK?

**Review by ToasterBiome — CHANGES_REQUESTED** [review]

File: code/game/objects/items/stunbaton.dm (hunk 9 above)
Comment (JamieD1) [reply code/game/objects/items/stunbaton.dm:55]: nope

File: code/__DEFINES/components.dm [hunk 10]
```diff
@@ -156,7 +156,7 @@
 #define COMSIG_MOVABLE_BUCKLE "buckle"							//from base of atom/movable/buckle_mob(): (mob, force)
 #define COMSIG_MOVABLE_UNBUCKLE "unbuckle"						//from base of atom/movable/unbuckle_mob(): (mob, force)
 #define COMSIG_MOVABLE_PRE_THROW "movable_pre_throw"			//from base of atom/movable/throw_at(): (list/args)
-#define COMSIG_MOVABLE_PRE_DROPTHROW "movable_pre_dropthrow"	//from base of atom/movable/throw_at(): (list/args)
+#define COMSIG_MOVABLE_PRE_DROPTHROW "movable_pre_dropthrow"	//from base of atom/movable/throw_at(): Triggers before throw just before the item is dropped
```
Comment (ToasterBiome) [inline code/__DEFINES/components.dm]: Is this before or after COMSIG_MOVABLE_PRE_THROW?

**Review by ToasterBiome — CHANGES_REQUESTED** [review]

File: code/__DEFINES/components.dm (hunk 10 above)
Comment (JamieD1) [reply code/__DEFINES/components.dm]: Before

Comment (ToasterBiome) [reply code/__DEFINES/components.dm]: Move it above it then and note it

Comment (JamieD1) [reply code/__DEFINES/components.dm]: Done

File: code/__DEFINES/components.dm [hunk 11]
```diff
@@ -155,6 +155,7 @@
 #define COMSIG_MOVABLE_IMPACT_ZONE "item_impact_zone"			//from base of mob/living/hitby(): (mob/living/target, hit_zone)
 #define COMSIG_MOVABLE_BUCKLE "buckle"							//from base of atom/movable/buckle_mob(): (mob, force)
 #define COMSIG_MOVABLE_UNBUCKLE "unbuckle"						//from base of atom/movable/unbuckle_mob(): (mob, force)
+#define COMSIG_MOVABLE_PRE_DROPTHROW "movable_pre_dropthrow"	//from base of atom/movable/throw_at(): Triggers before throw just before the item is dropped and before Pre Throw
```
Comment (ToasterBiome) [inline code/__DEFINES/components.dm]:
```suggestion
#define COMSIG_MOVABLE_PRE_DROPTHROW "movable_pre_dropthrow"	//from base of atom/movable/throw_at(): Triggers when throwing, before the item is dropped and before COMSIG_MOVABLE_PRE_THROW 
```

**Review by ToasterBiome — CHANGES_REQUESTED** [review]

**Review by ToasterBiome — APPROVED** [review]
this is only approval on the stuff i commented on

**Comment by ghost** [comment]
Literaly merging a PR with 9 downvotes

**Pull request merged**